From a636428c13f50df399652b8bf96cc3b4a0fb5602 Mon Sep 17 00:00:00 2001
From: Harshitha Prem <hprem@qti.qualcomm.com>
Date: Sun, 27 Jul 2025 22:06:45 +0530
Subject: [PATCH 8/8] QSDK: CP: hostapd: ATF: Add airtime distribution logic

Distribute airtime to all peers and ssid/ssid group
based on the configuration.

The distribution can be in 3 ways:
	1. when ssid config is enabled.
	2. when ssid_group is enabled.
	3. default, when dynamic atf is enabled (i.e., no ssid/ssid
group configuration)

Candidate list assumptions:
	1. In case of ssid config, each configured ssid will have
	   its own group and unconfigured ssids will be in default group.
	2. In case of ssid group, one or more ssid will be a group,
	   if unconfigured they will be in default group.

So, in both ssid config/ssid group case, group will be the entity on
which algorithm runs.

The following would be the distribution algorithm:
	1. If no peer is associated, no need for any distribution. Skip.
	2. For each group except default group, calculate for all the peer associated with
	   the group:
		a. configured peers calculation
			peer_airtime = ((group_airtime *
					 peer_configured_airtime)/ 100)
		b. For unconfigured peers:

			peer_airtime_calculation = residual_group_airtime/
						   no_of_unconfigured_peers

		Distribute equally the residual group airtime among unconfigured group clients
	3. Now assign the residual airtime equally to the default-group
peers.

Build the group config, peer config and wmm config expected by the
driver.

While building the peer config to driver consider below flags:
	A full update – Send about all the peers associated.
		1. When there is a reboot, and the first client gets associated.
		2. When atf_commit command is issued in between two command runs.
	A partial update – This will happen only in the following scenarios.
		The clients which get associated
		When a particular client gets disconnected and connected again.

Signed-off-by: Harshitha Prem <hprem@qti.qualcomm.com>
---
 src/atf/atf_offload.c | 343 +++++++++++++++++++++++++++++++++++++++++-
 src/atf/atf_offload.h |  62 +++++++-
 2 files changed, 397 insertions(+), 8 deletions(-)

--- a/src/atf/atf_offload.c
+++ b/src/atf/atf_offload.c
@@ -413,6 +413,7 @@ atf_allocate_algo()
 	dl_list_init(&algo->ssid_cfgs);
 	algo->num_peer_cfg = 0;
 	dl_list_init(&algo->peer_cfgs);
+	algo->init_update_done = 0;
 
 	return algo;
 }
@@ -441,10 +442,21 @@ atf_join_leave_update(struct hostapd_ifa
 		return;
 
 	ATF_SET_STA_TO_UPDATE(sta->atf_peer);
-	if (is_join)
+
+	/* After reboot, when first client joins it should be
+	 * full update.
+	 */
+	if (is_join && !iface->atf_algo->init_update_done) {
+		dl_list_init(&sta->atf_candidate_list);
+		ATF_OFFLOAD_SET_FULL_UPDATE(iface->atf_algo);
+		iface->atf_algo->init_update_done = 1;
+	} else if (is_join) {
+		dl_list_init(&sta->atf_candidate_list);
 		ATF_OFFLOAD_SET_JOIN_UPDATE(iface->atf_algo);
-	else
+	} else {
+		dl_list_del(&sta->atf_candidate_list);
 		ATF_OFFLOAD_SET_LEAVE_UPDATE(iface->atf_algo);
+	}
 
 	atf_trigger_config_timer(iface);
 }
@@ -472,8 +484,13 @@ atf_reset_groups(struct atf_algo *algo)
 	if (dl_list_empty(&algo->groups))
 		return;
 
-	dl_list_for_each(group, &algo->groups, struct atf_group, list)
+	dl_list_for_each(group, &algo->groups, struct atf_group, list) {
 		atf_reset_group_values(group);
+		if (os_strncmp(group->name, "default-group", strlen("default-group")) != 0) {
+			group->is_configured = 0;
+		}
+	}
+
 }
 
 
@@ -689,7 +706,7 @@ atf_update_peer(struct hostapd_data *hap
 				   MAC2STR(sta->addr));
 			return -1;
 		}
-
+		sta->atf_peer.group = group;
 		dl_list_add(&group->implicit_peers, &sta->atf_candidate_list);
 		group->num_impl_peers++;
 	} else {
@@ -701,6 +718,7 @@ atf_update_peer(struct hostapd_data *hap
 
 			return -1;
 		}
+		sta->atf_peer.group = group;
 		group->num_expl_peers++;
 	}
 
@@ -744,6 +762,7 @@ atf_update_peer_cfg_to_peer(struct atf_a
 
 	sta->atf_peer.atf_configured = true;
 	sta->atf_peer.peer_cfg_ref = peer_cfg;
+	sta->atf_peer.sta = sta;
 
 	/*update to the group candidate list*/
 	dl_list_add(&peer_cfg->group->explicit_peers, &sta->atf_candidate_list);
@@ -788,6 +807,320 @@ atf_build_candidate_list(struct hostapd_
 }
 
 
+void
+atf_cal_implicit_peers(struct atf_group *group)
+{
+	struct sta_info *sta;
+	u32 airtime;
+
+	if (dl_list_empty(&group->implicit_peers)) {
+		return;
+	}
+
+	if (group->num_impl_peers == 0)
+		return;
+
+	airtime = group->calculated_airtime / group->num_impl_peers;
+
+	dl_list_for_each(sta, &group->implicit_peers, struct sta_info, atf_candidate_list)
+		sta->atf_peer.calculated_airtime = airtime;
+}
+
+void
+atf_cal_explicit_peers(struct atf_group *group)
+{
+	struct sta_info *sta;
+	struct atf_peer_config *cfg = NULL;
+
+	if (dl_list_empty(&group->explicit_peers)) {
+		return;
+	}
+
+	dl_list_for_each(sta, &group->explicit_peers, struct sta_info, atf_candidate_list)
+	{
+		cfg = sta->atf_peer.peer_cfg_ref;
+
+		if (!cfg)
+			continue;
+
+		sta->atf_peer.calculated_airtime =
+		    ((group->user_cfg_airtime * cfg->user_cfg_airtime) /
+		     ATF_RADIO_DEFAULT_AIRTIME);
+		group->calculated_airtime -= sta->atf_peer.calculated_airtime;
+		group->total_explicit_airtime += sta->atf_peer.calculated_airtime;
+	}
+}
+
+int
+atf_distribute_airtime(struct hostapd_iface *iface)
+{
+	struct atf_algo *algo = iface->atf_algo;
+	struct atf_group *group, *def_group = NULL;
+	u16 iface_airtime = ATF_RADIO_DEFAULT_AIRTIME;
+
+	if (dl_list_empty(&algo->groups)) {
+		wpa_printf(MSG_ERROR, "ATF: group list is empty");
+		return -1;
+	}
+
+	dl_list_for_each(group, &algo->groups, struct atf_group, list)
+	{
+		if (!group->is_configured)
+			continue;
+
+		/* consider user configured group first and
+		 * residual airtime will be used by default-group.
+		 */
+		if (os_strncmp(group->name, "default-group",
+			       strlen("default-group")) == 0) {
+			def_group = group;
+		} else {
+			iface_airtime = iface_airtime - group->user_cfg_airtime;
+			group->calculated_airtime = group->user_cfg_airtime;
+			atf_cal_explicit_peers(group);
+			atf_cal_implicit_peers(group);
+		}
+	}
+
+	/*now calculate for the default group*/
+	group = def_group;
+	if (!group) {
+		wpa_printf(MSG_ERROR, "ATF: default-group missing");
+		return -1;
+	}
+
+	group->user_cfg_airtime = iface_airtime;
+	group->calculated_airtime = iface_airtime;
+	atf_cal_implicit_peers(group);
+	return 0;
+}
+
+int
+atf_offload_build_peer_config(struct hostapd_iface *iface,
+                              struct atf_peer_params *peer_param)
+{
+	int i;
+	struct atf_peer_info *peer_info;
+	u16 num_peers = 0;
+	struct atf_group *group = NULL;
+	struct atf_algo *algo = NULL;
+	struct hostapd_data *hapd;
+	struct sta_info *sta;
+
+	if (!iface->atf_algo) {
+		wpa_printf(MSG_ERROR, "ATF: algo is null");
+		return -1;
+	}
+	algo = iface->atf_algo;
+
+	if (algo->no_of_peers == 0) {
+		wpa_printf(MSG_ERROR, "ATF: There is no peer details to be sent to "
+		                      "driver. Skip build config.");
+		return 0;
+	}
+
+	peer_info = os_zalloc(algo->no_of_peers * sizeof(struct atf_peer_info));
+	if (!peer_info) {
+		wpa_printf(MSG_ERROR, "ATF: could not allocate peer info");
+		return -1;
+	}
+
+	for (i = 0; i < iface->num_bss; i++) {
+		hapd = iface->bss[i];
+		if (!hapd->started)
+			continue;
+
+		for (sta = hapd->sta_list; sta; sta = sta->next) {
+			if (sta && ap_sta_is_authorized(sta)) {
+
+				group = sta->atf_peer.group;
+
+				if (ATF_OFFLOAD_IS_LEAVE_UPDATE(algo) ||
+				    (ATF_OFFLOAD_IS_JOIN_UPDATE(algo) &&
+				     !ATF_IS_STA_UPDATED(sta->atf_peer)))
+					continue;
+
+				ATF_CLEAR_STA_UPDATED(sta->atf_peer);
+				memcpy(peer_info[num_peers].peer_macaddr, sta->addr, 6);
+				peer_info[num_peers].percentage_peer =
+				    sta->atf_peer.calculated_airtime;
+				peer_info[num_peers].group_index = group->index;
+
+				if (sta->atf_peer.atf_configured)
+					peer_info[num_peers].explicit_peer_flag = 1;
+
+				wpa_printf(MSG_DEBUG, "ATF: build peer " MACSTR " airtime %d group id %d %d",
+					   MAC2STR(sta->addr), sta->atf_peer.calculated_airtime, group->index,
+					   peer_info[num_peers].explicit_peer_flag);
+				num_peers++;
+			}
+		}
+	}
+
+	if (ATF_OFFLOAD_IS_FULL_UPDATE(algo) && algo->no_of_peers != num_peers) {
+		wpa_printf(MSG_ERROR,
+		           "ATF: Mismatched: Packed peers %d and no_of_peers %d",
+		           num_peers, algo->no_of_peers);
+		goto err_cleanup;
+	}
+
+	peer_param->full_update_flag = ATF_OFFLOAD_IS_FULL_UPDATE(algo);
+	peer_param->num_peers = num_peers;
+	peer_param->peer_info = peer_info;
+	wpa_printf(MSG_INFO, "ATF: build peer with %d peers and flags %d ", num_peers, peer_param->full_update_flag);
+
+	return 0;
+
+err_cleanup:
+	os_free(peer_info);
+	peer_info = NULL;
+	return -1;
+}
+
+int
+atf_offload_build_group_config(struct hostapd_iface *iface,
+                               struct atf_group_params *group_param)
+{
+	struct atf_group_param_info *group_info = NULL;
+	struct atf_algo *algo;
+	struct atf_group *group;
+	u8 i;
+
+	if (!iface->atf_algo)
+		return -1;
+
+	algo = iface->atf_algo;
+
+	/* Fill Group Configurations */
+	if (algo->num_group_cfg >= ATF_MAX_SSID_GROUP) {
+		wpa_printf(MSG_ERROR, "ATF: invalid num of Configured group %d!",
+		           algo->num_group_cfg);
+		return -1;
+	}
+
+	if (dl_list_empty(&algo->groups)) {
+		wpa_printf(MSG_ERROR, "ATF: group list is empty");
+		return -1;
+	}
+
+	group_info = os_zalloc(algo->num_group_cfg * sizeof(struct atf_group_param_info));
+	if (!group_info) {
+		wpa_printf(MSG_ERROR, "ATF: could not  allocate group info");
+		return -1;
+	}
+
+	i = 0;
+	dl_list_for_each(group, &algo->groups, struct atf_group, list)
+	{
+		if (!group->is_configured)
+			continue;
+
+		group_info[i].group_index  = group->index;
+		group_info[i].group_airtime = group->user_cfg_airtime;
+		group_info[i].total_implicit_peers = group->num_impl_peers;
+		group_info[i].total_explicit_peers = group->num_expl_peers;
+		group_info[i].group_policy = group->sched_policy;
+
+		/* Calculate total implicit peer units by removing total
+		 * explicit peer units from Group units.
+		 */
+		if (group->user_cfg_airtime >= group->total_explicit_airtime)
+			group_info[i].total_implicit_peer_units =
+				    (group->user_cfg_airtime -
+				     group->total_explicit_airtime);
+
+		wpa_printf(MSG_INFO,
+			   "ATF: Group id:%d airtime:%u  scheduling policy:%u "
+			    "unconfigured peers:%d configured_peers:%d "
+			    "implicit_peer_units:%d\n",
+			    group_info[i].group_index, group_info[i].group_airtime,
+			    group_info[i].group_policy,
+			    group_info[i].total_implicit_peers,
+			    group_info[i].total_explicit_peers,
+			    group_info[i].total_implicit_peer_units);
+		i++;
+	}
+	group_param->group_info = group_info;
+	group_param->num_groups = i;
+
+	return 0;
+}
+
+int
+atf_offload_build_wmm_ac_config(struct hostapd_iface *iface,
+                                struct atf_group_wmm_ac_params *wmm_ac_param)
+{
+	struct atf_group_wmm_ac_config *wmm_ac_cfg;
+	struct atf_algo *algo = iface->atf_algo;
+
+	if (dl_list_empty(&algo->groups)) {
+		wpa_printf(MSG_ERROR, "ATF: %s group list is empty", __func__);
+		return -1;
+	}
+
+	wmm_ac_cfg = os_zalloc(algo->num_group_cfg * sizeof(*wmm_ac_cfg));
+	if (!wmm_ac_cfg) {
+		wpa_printf(MSG_ERROR, "ATF: could not allocate wmm_ac_config");
+		return -1;
+	}
+
+	wmm_ac_param->num_groups = algo->num_group_cfg;
+
+	/* TODO: Logic to update all the ac values in future */
+
+	wmm_ac_param->wmm_ac_cfg = wmm_ac_cfg;
+	return 0;
+}
+
+int
+atf_send_calculated_airtime(struct hostapd_iface *iface)
+{
+
+	struct atf_peer_params atf_peer_param = {0};
+	struct atf_group_params atf_group_param = {0};
+	struct atf_group_wmm_ac_params atf_group_ac = {0};
+	int ret = -1;
+
+	if (!iface->atf_algo) {
+		wpa_printf(MSG_ERROR, "ATF: iface does not have atf_algo assigned");
+		return -1;
+	}
+
+	ret = atf_offload_build_peer_config(iface, &atf_peer_param);
+	if (ret != 0)
+		return ret;
+
+	/* Fill WMM AC configurations per group only in full Update mode */
+	if (ATF_OFFLOAD_IS_FULL_UPDATE(iface->atf_algo)) {
+		ret = atf_offload_build_wmm_ac_config(iface, &atf_group_ac);
+		if (ret != 0) {
+			goto peer_param_free;
+		}
+	}
+
+	ret = atf_offload_build_group_config(iface, &atf_group_param);
+	if (ret != 0) {
+		goto group_ac_free;
+	}
+
+	/* TODO: send the built params to driver */
+
+	ATF_OFFLOAD_SET_NO_UPDATE(iface->atf_algo);
+
+	if (atf_group_param.group_info)
+		os_free(atf_group_param.group_info);
+
+group_ac_free:
+	if (atf_group_ac.wmm_ac_cfg)
+		os_free(atf_group_ac.wmm_ac_cfg);
+
+peer_param_free:
+	if (atf_peer_param.peer_info)
+		os_free(atf_peer_param.peer_info);
+
+	return ret;
+}
+
 static void
 atf_cfg_timeout_handler(void *eloop_ctx, void *timeout_ctx)
 {
@@ -819,6 +1152,14 @@ atf_cfg_timeout_handler(void *eloop_ctx,
 		goto out;
 	}
 
+	if (atf_distribute_airtime(iface) != 0) {
+		wpa_printf(MSG_ERROR, "ATF: could not distribute airtime");
+		goto out;
+	}
+
+	if (atf_send_calculated_airtime(iface) != 0)
+		goto out;
+
 out:
 	algo->atf_tasksched = 0;
 	return;
--- a/src/atf/atf_offload.h
+++ b/src/atf/atf_offload.h
@@ -23,6 +23,11 @@
 #define ATF_MAX_SSID 16
 #define ATF_MAX_PEER 512
 
+/* Percentage value scaled by 10 to avoid decimals. (e.g., 10.5% → 105)
+ * So, one digit after the decimal is accounted and represented as integer.
+ */
+#define ATF_RADIO_DEFAULT_AIRTIME 1000
+
 /**
  * enum atf_offload_update_flag - Airtime update flags for driver
  * @ATF_OFFLOAD_FULL_UPDATE: Full update and default update method
@@ -90,6 +95,7 @@ struct atf_peer_config {
 
 struct atf_peer {
 	bool atf_configured;
+	struct atf_group *group;
 	struct atf_peer_config *peer_cfg_ref;
 	struct sta_info *sta;
 	u32 calculated_airtime;
@@ -125,6 +131,7 @@ struct atf_group {
 	u32 num_of_ssid;
 	char ssidname[WLAN_SSID_MAX][WLAN_SSID_MAX_LEN + 1];
 	u32 user_cfg_airtime;
+	u32 sched_policy;
 
 	/* add this node to atf_algo */
 	struct dl_list list;
@@ -169,6 +176,12 @@ struct atf_algo {
 	/*Flag to indicate update task scheduled */
 	bool atf_tasksched;
 
+	/* first sta association should be sent as
+	 * full update. This flag represents whether
+	 * it is done or not
+	 */
+	bool init_update_done;
+
 	/* no. of peers to be updated to driver*/
 	u16 no_of_peers;
 
@@ -185,6 +198,55 @@ struct atf_offload {
 	struct dl_list algo_list;
 };
 
+
+struct atf_peer_info {
+	u8 peer_macaddr[6];
+	u16 percentage_peer;
+	u16 group_index;
+	u32 explicit_peer_flag;
+};
+
+struct atf_peer_params {
+	u16 num_peers;
+	u8 full_update_flag;
+	struct atf_peer_info *peer_info;
+};
+
+struct atf_group_param_info {
+	u16 group_index;
+	u16 group_airtime;
+	u16 total_implicit_peers;
+	u16 total_explicit_peers;
+	u16 total_implicit_peer_units;
+	u32 group_policy;
+};
+
+struct atf_group_params {
+	u8 num_groups;
+	struct atf_group_param_info *group_info;
+};
+
+struct atf_group_wmm_ac_config {
+	u32 ac_be; /* Relative ATF % for BE */
+	u32 ac_bk; /* Relative ATF % for BK */
+	u32 ac_vi; /* Relative ATF % for VI */
+	u32 ac_vo; /* Relative ATF % for VO */
+	u32 reserved[2];
+};
+
+struct atf_group_wmm_ac_params {
+	u32 num_groups;
+	struct atf_group_wmm_ac_config *wmm_ac_cfg;
+};
+
+struct atf_group_info {
+	u8 percentage_group;
+	u8 implicit_peer;
+	u8 explicit_peers;
+	u8 implicit_peer_units;
+	u8 group_units_reserved;
+};
+
 #ifdef CONFIG_ATF_OFFLOAD
 
 void atf_offload_init(struct hapd_interfaces *ifaces);
